[lldb] Add DWARFExpressionEntry and GetExpressionEntryAtAddress() to …

UltimateForce21 opened the pull request:
This patch introduces a new struct and helper API in `DWARFExpressionList` to expose 
variable location metadata (base address, end address, and DWARFExpression pointer) for 
a given PC address. It will be used in later patches to annotate disassembly instructions 
with source-level variable locations.

## New struct
```
/// Represents one entry in a DWARFExpressionList, with its range and expr.
struct DWARFExpressionEntry {
  lldb::addr_t base;           // file‐address start of this location range
  lldb::addr_t end;            // file‐address end of this range (exclusive)
  const DWARFExpression *expr; // the DWARF expression for this range
};
```

## New API
```
/// Retrieve the DWARFExpressionEntry covering a particular instruction.
///
/// \param func_load_addr
///     The load address of the start of the function containing this location list;
///     used to translate between file offsets and load addresses. If this is
///     LLDB_INVALID_ADDRESS, the stored CU base (m_func_file_addr) is used.
///
/// \param load_addr
///     The load address of the *current* PC (i.e., the instruction for which
///     we want its variable‐location entry). We first convert this back into
///     the function’s file‐address space to find the correct DWARF range.
///
/// \returns
///     On success, an entry whose `[base,end)` covers this PC; else an Error.
llvm::Expected<DWARFExpressionEntry>
GetExpressionEntryAtAddress(lldb::addr_t func_load_addr,
                            lldb::addr_t load_addr) const;
```

## Rationale
LLDB already provides:
```
const DWARFExpression *
GetExpressionAtAddress(lldb::addr_t func_load_addr,
                       lldb::addr_t load_addr) const;

```
However, this only returns the DWARF expression itself, without the file‐address start (base) and end (end) of the location range. Those bounds are crucial for:

1) Detecting range beginnings: render a var = <location> annotation exactly when a variable’s live‐range starts.

2) Detecting range continuation: optionally display a “|” on subsequent instructions in the same range.

3) Detecting state changes: know when a variable moves (e.g. from one register to another), becomes a constant, or goes out of scope.

These primitives form the foundation for the Rich Disassembler feature proposed for GSoC 25.

UltimateForce21 committed:
[lldb] Add DWARFExpressionEntry and GetExpressionEntryAtAddress() to DWARFExpressionList

This introduces a new API for retrieving DWARF expression metadata associated with variable location entries at a given PC address. It provides the base, end, and expression pointer for downstream consumers such as disassembler annotations.

Intended for use in richer instruction annotations in Instruction::Dump().

UltimateForce21 committed:
Update lldb/include/lldb/Expression/DWARFExpressionList.h

Co-authored-by: Jonas Devlieghere <jonas@devlieghere.com>

UltimateForce21 committed:
Update lldb/include/lldb/Expression/DWARFExpressionList.h

Updated comment for GetExpressionEntryAtAddress to directly refer to struct DWARFExpressionEntry

Co-authored-by: Jonas Devlieghere <jonas@devlieghere.com>

UltimateForce21 committed:
Update lldb/source/Expression/DWARFExpressionList.cpp

updating code style for function GetExpressionEntryAtAddress

Co-authored-by: Jonas Devlieghere <jonas@devlieghere.com>

UltimateForce21 commented on lldb/include/lldb/Expression/DWARFExpressionList.h:
```diff
@@ -59,6 +59,18 @@ class DWARFExpressionList {
 
+  /// Represents an entry in the DWARFExpressionList with all needed metadata
+  struct DWARFExpressionEntry {
+    lldb::addr_t base;
+    lldb::addr_t end;
```
These are file addresses (without offsets applied), I will try to make that more clear both in the comments and in the variable names. I'll also take a look into the AddressRange class and see if I can utilize it, thank you for the pointers.

UltimateForce21 commented on lldb/source/Expression/DWARFExpressionList.cpp:
```diff
@@ -53,6 +53,27 @@ bool DWARFExpressionList::ContainsAddress(lldb::addr_t func_load_addr,
+  uint32_t index = m_exprs.FindEntryIndexThatContains(addr);
+  if (index == UINT32_MAX) {
+    return llvm::createStringError(llvm::inconvertibleErrorCode(),
+                                   "No DWARF expression found for address 0x%llx", addr);
```
Sounds good, I will refactor it to return std::optional instead.

UltimateForce21 commented on lldb/source/Expression/DWARFExpressionList.cpp:
```diff
@@ -53,6 +53,27 @@ bool DWARFExpressionList::ContainsAddress(lldb::addr_t func_load_addr,
+}
+
 const DWARFExpression *
 DWARFExpressionList::GetExpressionAtAddress(lldb::addr_t func_load_addr,
```
As you know the new function GetExpressionEntryAtAddress, does almost exact same thing as this one. I don't see this function being used anywhere else in the lldb code base, so maybe it can be removed, but to be safe I did not want to alter it. But from what I can tell, it should be okay to remove it. We can also choose to keep it and maybe refactor it to just call the GetExpressionEntryAtAddress and just return the `DWARFExpression *expr`.

adrian-prantl replied:
If it's completely unused, and there is no indication that was added in support of any downstream clients (either as a comment or in the commit message) we could consider deleting it in a separate PR. Not urgent.

UltimateForce21 committed:
Update DWARFExpressionList.h

Replace raw base/end with `AddressRange` in `DWARFExpressionEntry` and cleans up helper comments to follow Doxygen convention.

Using `AddressRange` makes the intent clearer, avoids duplication of basic `AddressRange` logic usage

UltimateForce21 committed:
Update DWARFExpressionList.cpp

Converts `GetExpressionEntryAtAddress` to return `llvm::Expected<DWARFExpressionEntry>` using the updated `DWARFExpressionEntry`. Updates the implementation to compute a single `AddressRange file_range` for each DWARF location interval.

UltimateForce21 committed:
Update DWARFExpressionList.h

Updated commenting style for struct DWARFExpressionEntry

UltimateForce21 committed:
Update DWARFExpressionList.cpp

Updated function `llvm::Expected<DWARFExpressionList::DWARFExpressionEntry>
DWARFExpressionList::GetExpressionEntryAtAddress` to use `FindEntryThatContains` instead of `FindEntryIndexThatContains`

UltimateForce21 committed:
Change GetExpressionEntryAtAddress to return std::optional instead of llvm::Expected

adrian-prantl commented on lldb/source/Expression/DWARFExpressionList.cpp:
```diff
@@ -53,6 +54,30 @@ bool DWARFExpressionList::ContainsAddress(lldb::addr_t func_load_addr,
+    func_load_addr = m_func_file_addr;
+  
+  // translate to file-relative PC
+  lldb::addr_t file_pc = load_addr - func_load_addr + m_func_file_addr;
```
You should be using subOverflow (https://llvm.org/doxygen/MathExtras_8h.html) here to guard against wrap-around.

UltimateForce21 replied:
I dug into MathExtras and found that its AddOverflow and SubOverflow helpers are only defined for signed types. I they rely on two’s-complement signed overflow semantics and return the truncated signed result. Since lldb::addr_t is an unsigned address, I think using those helpers would either fail to compile or silently convert to signed semantics. So I tried to implement the signed overflow check into the function. Please let me know if you think it is alright.

adrian-prantl reviewed (commented):
LGTM with under/overflow checks added.

UltimateForce21 committed:
Update DWARFExpressionList.cpp

Co-authored-by: Adrian Prantl <adrian.prantl@gmail.com>

UltimateForce21 committed:
Add underflow/overflow checks to GetExpressionEntryAtAddressi
This patch adds explicit checks:
  - ensure `load_addr >= func_load_addr` to avoid underflow,
  - compute and verify a temporary delta variable, then verify `delta + m_func_file_addr` does
    not exceed `addr_t` max to avoid overflow.

UltimateForce21 commented on lldb/source/Expression/DWARFExpressionList.cpp:
```diff
@@ -53,6 +54,38 @@ bool DWARFExpressionList::ContainsAddress(lldb::addr_t func_load_addr,
+DWARFExpressionList::GetExpressionEntryAtAddress(lldb::addr_t func_load_addr,
+                                                lldb::addr_t load_addr) const {
+  if (const DWARFExpression *always = GetAlwaysValidExpr()) {
+    AddressRange full_range(m_func_file_addr, /*size=*/LLDB_INVALID_ADDRESS);
```
I believe you are right it is safer and probably more clear for users, as well.

UltimateForce21 replied:
I believe using an optional `AddressRange` the way you said is a safer option and probably more clear for users, I have committed the change.

UltimateForce21 committed:
Make file_range optional in DWARFExpressionEntry for always-valid expr

JDevlieghere reviewed (approved):
LGTM

UltimateForce21 committed:
Merge branch 'main' into add-dwarfexprentry-api

UltimateForce21 committed:
Merge branch 'main' into add-dwarfexprentry-api

UltimateForce21 committed:
Fixed formating issues
